athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -0,0 +1,282 @@
+            }
+            catch (IOException error) {
+                logger.error("Error getting indices:", error);
+            }
+            finally {
+                client.shutdown();
+            }
+        }
+        else {
+            logger.warn("Domain does not exist in map: " + request.getSchemaName());
+        }
+
+        return new ListTablesResponse(request.getCatalogName(), indices);
+    }
+
+    /**
+     * Used to get definition (field names, types, descriptions, etc...) of a Table.
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request Provides details on who made the request and which Athena catalog, database, and table they are querying.
+     * @return A GetTableResponse which primarily contains:
+     * 1. An Apache Arrow Schema object describing the table's columns, types, and descriptions.
+     * 2. A Set<String> of partition column names (or empty if the table isn't partitioned).
+     * 3. A TableName object confirming the schema and table name the response is for.
+     * 4. A catalog name corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public GetTableResponse doGetTable(BlockAllocator allocator, GetTableRequest request)
+    {
+        logger.info("doGetTable: enter - " + request);
+
+        Schema schema = null;
+
+        // Look at GLUE catalog first.
+        try {
+            if (awsGlue != null) {
+                schema = super.doGetTable(allocator, request).getSchema();
+                logger.info("doGetTable: Retrieved schema for table[{}] from AWS Glue.", request.getTableName());
+            }
+        }
+        catch (Exception error) {
+            logger.warn("doGetTable: Unable to retrieve table[{}:{}] from AWS Glue.",
+                    request.getTableName().getSchemaName(),
+                    request.getTableName().getTableName(),
+                    error);
+        }
+
+        // Supplement GLUE catalog if not present.
+        if (schema == null) {
+            String endpoint = ElasticsearchHelper.getDomainEndpoint(request.getTableName().getSchemaName());
+
+            if (!endpoint.isEmpty()) {
+                String index = request.getTableName().getTableName();
+                AwsRestHighLevelClient client = clientFactory.getClient(endpoint);
+                try {
+                    LinkedHashMap<String, Object> mappings = client.getMapping(index);
+                    LinkedHashMap<String, Object> meta = new LinkedHashMap<>();
+
+                    if (mappings.containsKey("_meta")) {
+                        meta.putAll((LinkedHashMap) mappings.get("_meta"));
+                    }
+
+                    schema = ElasticsearchHelper.parseMapping(mappings, meta);
+                }
+                catch (IOException error) {
+                    logger.error("Error mapping index:", error);
+                }
+                finally {
+                    client.shutdown();
```

Comment:
possible to use auto-closeable for this? also do we need to close the client each time or can it be reused?

Reply:
We discussed this before. Most of the Elasticsearch methods are declared final (including close) which makes them not usable with Mockito much to my chagrin (unless we upgrade to version 2). With regard to client reusability, I'm all for it as long as you can give me an alternative place to close the client. From the elasticsearch docs: "It is important that it gets closed when no longer needed so that all the resources used by it get properly released, as well as the underlying http client instance and its threads."